src/index/base.cpp

```diff
@@ -262,6 +271,14 @@ void BaseIndex::BlockConnected(const std::shared_ptr<const CBlock>& block, const
     }
 
     if (WriteBlock(*block, pindex)) {
+        // TODO assumeutxo: since blocks may be submitted to this
+        // indexer out of order (i.e. earlier blocks from the background
+        // validation chainstate submitted *after* more recent blocks from the
+        // snapshot chainstate), look at only updating this when pindex's
+        // height is greater than best_block's.
+        //
+        // Note: I tried this earlier, but got some test failures
+        // (functional/rpc_rawtransactions.py and rpc_getblockfilter.py).
         m_best_block_index = pindex;
```

Comment:
I think if anything, probably the reverse; if we only do this accounting for the active chain, we run the risk of giving the impression that the index has been fully built up until that point (when in reality we're waiting for block data to come in further back in the chain via the background chainstate).

This requires more thought, but my inclination is to either (i) explicitly reform the assumption that indexes will be built sequentially, or (ii) disallow `BlockConnected` handling for the active chainstate when there is a background chainstate in use.

I'll add plenty of documentation around this in the next day or so. And I should probably try to write some tests for the interaction between indexation and assumeutxo use somehow.

Reply:
Although it's worth noting that `BlockConnected` signals are [ignored by the indexer](https://github.com/jamesob/bitcoin/blob/bd07c1bc78ca41d7f03648ccf0c0db102bde9cc9/src/index/base.cpp#L242-L244) until a full indexation has completed, so this may be a moot point anyway (which I think is why I didn't materially change the code in the first place...).

Reply:
> Although it's worth noting that BlockConnected signals are ignored by the indexer until a full indexation has completed

Hah, just kidding - turns out that `m_synced` flips to `true` in a degenerate case when we start up with an empty datadir; i.e. if a user enables `-txindex=1` on first use. I'm not sure if jimpo intended this to happen when he wrote it, but because `m_chain` has no entries the latch flips immediately. So the above discussion is not moot, and this (maybe surprising) behavior should probably be documented in `BaseIndex`. I may do that in a separate PR soon.